### DISCUSSION TITLE
Best way to to pass data to service so it can be re-used?

### DISCUSSION THREAD

**Discussion by jmlivingston** [statelyai/xstate#2955] [Q&A] [answered]
What is the best way pass data to a service so it can be re-used? This isn’t a very practical example, but I’m trying to simplify a more complicated one.

- Add a key (dataValue) to the return value or the invoke src and assign accordingly - [https://codesandbox.io/s/use-same-action-data-from-service-ii5zq](https://codesandbox.io/s/use-same-action-data-from-service-ii5zq)

- Use a meta data property (dataValue). I’d really like to do it this way, but this feels code feels like a hack. - [https://codesandbox.io/s/use-same-action-data-from-meta-x7igr](https://codesandbox.io/s/use-same-action-data-from-meta-x7igr)

`actionMeta.state.meta[`(machine).${actionMeta.state.value}`].dataValue`

- Just use a different service and assign accordingly - [https://codesandbox.io/s/use-separate-service-szt2f]( https://codesandbox.io/s/use-separate-service-szt2f)

> Note: I've been watching xstate for a couple years, but just starting to use xstate. Very excited about the potential.

**Comment by davidkpiano** [comment, marked as answer]
I would prefer using a factory function instead:

```js
function createSomeInvoke(location) {
  return {
    src: /* ... */,
    // ...
    onDone: {
      actions: assign({ [location]: (ctx, e) => e.data })
    }
  }
}

// ...


one: {
  invoke: createSomeInvoke('one')
},
two: {
  invoke: createSomeInvoke('two')
}
```

What do you think?

**Reply by jmlivingston** [reply]
Thanks a bunch for the feedback! I think this will do the trick.
